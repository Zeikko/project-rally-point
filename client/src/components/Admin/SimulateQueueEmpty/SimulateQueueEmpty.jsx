import _ from 'lodash'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { leaveGameAction } from '../../../actions/player-actions'
import Button from '../../Button/Button'
import * as propTypes from '../../../constants/prop-types'

class SimulateQueueEmpty extends Component {
  constructor() {
    super()
    this.handleClick = this.handleClick.bind(this)
  }

  async handleClick() {
    const { dispatch, game } = this.props
    for(let id = 1; id <= 48; id++) {
      dispatch(leaveGameAction(game.data.id, id))
      await sleep(_.random(10, 100))
    }
  }

  render() {
    return (
      <Button onClick={this.handleClick}>
        Simulate Queue Empty
      </Button>
    )
  }
}

SimulateQueueEmpty.propTypes = {
  dispatch: PropTypes.func.isRequired,
  game: propTypes.game.isRequired,
}

export default SimulateQueueEmpty

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
