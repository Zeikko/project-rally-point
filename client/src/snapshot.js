import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}