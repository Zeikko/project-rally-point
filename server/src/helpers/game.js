import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'
import playerRoles from '../../../common/player-roles.json'

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadMemberPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfSquadLeaders = await db('player')
    .select('id')
    .where({
      gameId,
      role: playerRoles.SQUAD_LEADER,
      team: game.teamWithTurnToPick,
    })
  return (numberOfSquadLeaders.length + 1) === getOptimalNumberOfSquads(game)
}

export async function shouldStartPlayingGame(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
    .whereNotNull('team')
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_LEADER_PICK,
      teamWithTurnToPick: 1,
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadMemberPick(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.SQUAD_MEMBER_PICK })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startPlayingGame(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.PLAYING })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

export async function passSquadLeaderPickTurn(game) {
  const games = await db('game')
    .update({
      teamWithTurnToPick: (game.teamWithTurnToPick % 2) + 1,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

export async function passPlayerPickTurn(game, playersCount) {
  const nextTeamWithTurnToPick = await calculatePickTurn(game, playersCount)
  const games = await db('game')
    .update({
      teamWithTurnToPick: nextTeamWithTurnToPick,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

async function calculatePickTurn(game, playersCount) {
  const numberOfSquads = getOptimalNumberOfSquads(game)
  if (playersCount % numberOfSquads === 0) {
    return (game.teamWithTurnToPick % 2) + 1
  }
  return game.teamWithTurnToPick
}

export function getOptimalNumberOfSquads(game) {
  return Math.ceil(game.maxPlayers / 2 / 9)
}
