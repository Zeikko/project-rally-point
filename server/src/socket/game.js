import _ from 'lodash'
import db from '../db'
import actions from '../../../common/actions.json'

export default function handleGame(io, socket, action) {
  switch (action.type) {
    case actions.GET_GAME_REQUEST:
      return getGameRequest(io, socket, action)
    default:
      return null
  }
}

export async function getGameRequest(io, socket) {
  try {
    const game = await db('game')
      .first('*')
      .where({
        status: 'queue',
      })
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
  } catch (error) {
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
}

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: 'voting captains' })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function createNewGame() {
  return db('game').insert({})
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

