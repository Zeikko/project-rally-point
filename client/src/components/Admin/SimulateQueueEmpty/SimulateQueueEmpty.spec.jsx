import React from 'react'
import { mount } from 'enzyme'
import { matchComponentToSnapshot } from '../../../test/snapshot'
import SimulateQueueEmpty from './SimulateQueueEmpty'
import game from '../../../../../fixtures/game.json'

describe('SimulateQueueEmpty', () => {
  it('renders correctly', () => {
    matchComponentToSnapshot(<SimulateQueueEmpty dispatch={() => {}} gameState={{ game }} />)
  })

  it('joins a game', () => {
    const dispatch = jest.fn()
    const wrapper = mount(<SimulateQueueEmpty dispatch={dispatch} gameState={{ game }} />)
    wrapper.find('button').simulate('click')
    expect(dispatch.mock.calls[0]).toEqual([
      {
        type: 'LEAVE_GAME_REQUEST',
        socket: true,
        gameId: 1,
        userId: 1,
      },
    ])
  })
})
