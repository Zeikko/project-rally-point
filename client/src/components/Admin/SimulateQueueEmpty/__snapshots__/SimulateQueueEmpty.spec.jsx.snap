// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulateQueueEmpty renders correctly 1`] = `
.c0 {
  font-size: 18px;
  border: solid black 1px;
  background-color: #AAA;
  cursor: pointer;
}

<button
  className="c0"
  onClick={[Function]}
>
  Simulate Queue Empty
</button>
`;
